stmtools/_io.py

```diff
@@ -0,0 +1,137 @@
+        2. Per space-time attribute, a common RE pattern is shared by all columns. E.g. for the 
+          time-sries of amplitude data, the column names are "amp_20100101", "amp_20100110", 
+          "amp_20100119" ..., where "^amp_" is the common RE pattern;
+        3. There is no temporal-only (i.e. 1-row attribute) attribute presented the csv file.
+
+    `from_csv` does not retrieve time stamps based on column names. The `time` coordinate of 
+      the output STM will be a monotonic integer series starting from 0.
+
+    Args:
+        file (str | Path): Path to the csv file.
+        space_pattern (str, optional): RE pattern to match space attribute columns. 
+          Defaults to "^pnt_".
+        spacetime_pattern (dict | None, optional): A dictionay mapping RE patterns of each 
+          space-time attribute to corresponding variable names. Defaults to None, which means
+          the following map will be applied: 
+          {"^d_": "deformation", "^a_": "amplitude", "^h2ph_": "h2ph"}.
+        coords_cols (list | dict, optional): List of columns to be used as space coordinates. 
+          When `coords_cols` is a dictionary, a reaming will be performed per coordinates. 
+          Defaults to None, then the following renaming will be performed: 
+          "{"pnt_lat": "lat", "pnt_lon": "lon"}"
+        output_chunksize (dict | None, optional): Chunksize of the output. Defaults to None, 
+          then the size of the first chunk in the DaskDataFrame will be used, up-rounding to 
+          the next 5000.
+        blocksize (int | str | None, optional): Blocksize to load the csv. 
+          Defaults to 200e6 (in bytes). See the documentation of 
+          [dask.dataframe.read_csv](https://docs.dask.org/en/stable/generated/dask.dataframe.read_csv.html)
+
+    Returns:
+        xr.Dataset: Output STM instance
+    """
+
+    # Load csv as Dask DataFrame
+    ddf = dd.read_csv(file, blocksize=blocksize)
+
+    # Take the first column and compute the chunk sizes
+    da_col0 = ddf[ddf.columns[0]].to_dask_array()
+    da_col0.compute_chunk_sizes()
+    chunks = da_col0.chunks
+
+    # Count time dimension by one spacetime column
+    if spacetime_pattern is None:
+        spacetime_pattern = {"^d_": "deformation", "^a_": "amplitude", "^h2ph_": "h2ph"}
+    time_shape = 0
+    if spacetime_pattern is not None:
+        key = list(spacetime_pattern.keys())[0]
+        for column in ddf.columns:
+            if re.match(re.compile(key), column):
+                time_shape += 1
+
+    # Initiate a template STM
+    coords = {
+        "space": range(da_col0.shape[0]),
+        "time": range(time_shape),
+    }
+    stmat = xr.Dataset(coords=coords)
+
+    # Read csv by columns
+    dict_temp_da = dict()  # Temporary dask array collector
+    for k in spacetime_pattern.keys():
+        dict_temp_da[k] = []
+
+    # Temporaly save time-series columns to lists in dict_temp_da
+    for column in ddf.columns:
```

Comment:
I am not sure whether reading a (large) CSV file column by column is a good idea, and I wonder how Dask does it internally. What I am afraid it could happen is that Dask is actually loading the full CSV into memory, or you are reading through all the rows multiple times and each time extract only one value. Probably it is better to read row-oriented formats like CSV files in blocks of rows.

Reply:
Good suggestion! I changed the implementation to fist convert the entire DaskDataframe to DaskArray, then index the DaskArray based on the column index

Reply:
Uhm, would this change how the data is read in? I think you would still be loading all the dataset in order to extract even a single column

Reply:
Performance-wise, I think it would work better if you would read one partition and load all columns from it, then read another partition, and load all columns again, and so on

Reply:
Ah I see what you mean. That make sense. I will think of a way how to do this

Reply:
Hi @fnattino, I inspected the following reading resullts:

```python
data = stmtools.from_csv(p_data, blocksize=100e6)
data
```

```
<xarray.Dataset>
Dimensions:                (space: 50000, time: 198)
Coordinates:
  * space                  (space) int64 0 1 2 3 4 ... 49996 49997 49998 49999
  * time                   (time) int64 0 1 2 3 4 5 ... 192 193 194 195 196 197
    lat                    (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    lon                    (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
Data variables: (12/25)
    pnt_id                 (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    pnt_flags              (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    ...                     ...
    pnt_std_linear         (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    pnt_std_quadratic      (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    pnt_std_seasonal       (space) object dask.array<chunksize=(25000,), meta=np.ndarray>
    deformation            (space, time) object dask.array<chunksize=(25000, 198), meta=np.ndarray>
    amplitude              (space, time) object dask.array<chunksize=(25000, 198), meta=np.ndarray>
    h2ph                   (space, time) object dask.array<chunksize=(25000, 198), meta=np.ndarray>
```

The data has a `space` size of 50000 and has a chunk size `25000`, so it should be chunked into 2 parts in space.  I visualized two dask graphs of the loaded Dataset, focusing on 1 and 2 data variables:

```python
# Select some vars and remove coords to simplify
data2_one_var = data[['pnt_id']].drop(['lat', 'lon'])
data2_two_var = data[['pnt_id', 'pnt_flags']].drop(['lat', 'lon'])
```

Graph for 1 variable:
![dask_1var](https://github.com/MotionbyLearning/stmtools/assets/33962195/4bddcd85-4712-4b04-8159-9b2275c37dab)

Graph for 2 variables: 
![dask_2var](https://github.com/MotionbyLearning/stmtools/assets/33962195/f257b454-98e6-423a-942a-c1ceb53b9ea0)

I interpret the two graphs as Dask is already doing what we want it to do: first loading each row (space) chunk, and within each chunk get the column attribute. Therefore I would suggest we leave the implementation like this? I will add some documentation to doc this behavior of Dask.

Reply:
Sounds good! Also, we can test how the function performs, and if we see it's becoming a bottleneck we can have a look at it again in future!